adr/0002-scripting-engine.md

```diff
@@ -0,0 +1,170 @@
+# 2. Scripting Engine
+
+Date: 2024-06-03
+
+## Status
+
+Draft
+
+## Context
+
+Presently, only action types provided by `maru`[^1] are:
+
+- `cmd:` (i.e. [`BaseAction`](https://github.com/defenseunicorns/maru-runner/blob/main/src/types/actions.go#L23)) - basic shell command execution
+- `wait:` (i.e. [`ActionWait`](https://github.com/defenseunicorns/maru-runner/blob/main/src/types/actions.go#L37)), which supports two types of "status checks":
+  - `cluster:` (i.e. [`ActionWaitCluster`](https://github.com/defenseunicorns/maru-runner/blob/main/src/types/actions.go#L43)) -
+    perform status checks against K8s cluster resources
+  - `network:` (i.e. [`ActionWaitNetwork`](https://github.com/defenseunicorns/maru-runner/blob/main/src/types/actions.go#L51)) -
+    poll arbitrary HTTP/TCP endpoints for a given status code
+
+The `wait` action is really the only abstraction that is provided around
+shell scripting. As we seek to enhance or expand built-in capabilities,
+we have three high-level options for doing so:
+
+1. expand YAML-based DSL with more configuration options
+2. vendor additional "tools" and encourage `./zarf <tool> [...]` pattern
+3. **provide a cross-platform scripting engine (with builtins
+   for common tasks)**
+
+Below is an example of a common use case for HTTP status checks that
+is not readily solved by the existing `wait.network:` probe:
+
+```yaml
+tasks:
+  - description: SonarQube UI Status Check
+    maxRetries: 30
+    cmd: |
+      STATUS=$(curl -s 'https://sonarqube.uds.dev/api/system/status' | ./uds zarf tools yq '.status')
+      echo "SonarQube system status: ${STATUS}"
+      if [ $STATUS != "UP" ]; then
+        sleep 10
+        exit 1
+      fi
+```
+
+### Native Scripting Engine
+
+For the purposes of this proposal, we will scope the evaluation to libraries
+that can be embedded natively.
+
+Tools like [`zx`](https://google.github.io/zx/getting-started)
+(which provides a JS API on top of shell scripting) are interesting, but
+anything that is not written in Go would be difficult to integrate in a way
+that actually improves portability of user-defined scripts.
```

Comment:
```suggestion
that actually improves portability of user-defined scripts without delegating to more complex paradigms like WASM.
```
(as Jeff Goldbloom says "life finds a way" and zx _could_ be embedded but it would be quite annoying to do so)

Reply:
> For the purposes of this proposal, we will scope the evaluation to libraries that can be embedded natively.

> anything that is not written in Go would be difficult to integrate

I think the point stands as is. `zx` could not be integrated using WASM either because it relies on the Node.JS runtime. Your statement presents WASM as "complex" but similar to a native runtime, which it is not. It is just portable bytecode.

My point here was to scope the eval to Go-native runtimes. Otherwise, as you say, "life finds a way" and literally anything could be considered.